File: mycuration/src/main/java/com/phicdy/mycuration/view/activity/FeedUrlHookActivity.java
package com.phicdy.mycuration.view.activity;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.annotation.StringRes;
import android.support.annotation.UiThread;
import android.widget.Toast;

import com.phicdy.mycuration.R;
import com.phicdy.mycuration.db.DatabaseAdapter;
import com.phicdy.mycuration.presenter.FeedUrlHookPresenter;
import com.phicdy.mycuration.rss.RssParser;
import com.phicdy.mycuration.rss.UnreadCountManager;
import com.phicdy.mycuration.task.NetworkTaskManager;
import com.phicdy.mycuration.tracker.GATrackerHelper;
import com.phicdy.mycuration.view.FeedUrlHookView;

import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

public class FeedUrlHookActivity extends Activity implements FeedUrlHookView {

    private FeedUrlHookPresenter presenter;
	private BroadcastReceiver receiver;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_feed_url_hook);

        DatabaseAdapter dbAdapter = DatabaseAdapter.getInstance();
        UnreadCountManager unreadCountManager = UnreadCountManager.getInstance();
        NetworkTaskManager networkTaskManager = NetworkTaskManager.INSTANCE;
        RssParser parser = new RssParser();
		presenter = new FeedUrlHookPresenter(dbAdapter, unreadCountManager, networkTaskManager, parser);
		presenter.setView(this);

		Intent intent = getIntent();
		final String action = intent.getAction();
        String url = null;
        if (action.equals(Intent.ACTION_VIEW)) {
            url = intent.getDataString();
        }else if (action.equals(Intent.ACTION_SEND)) {
            // For Chrome
            Bundle extras = getIntent().getExtras();
            CharSequence urlChar = extras.getCharSequence(Intent.EXTRA_TEXT);
            if (urlChar != null) {
                url = urlChar.toString();
            }
        }
        if (url != null) {
            presenter.handle(action, url);
        }
        GATrackerHelper.INSTANCE.sendScreen(getString(R.string.add_rss_from_intent));
	}

	@Override
	protected void onPause() {
		super.onPause();
        presenter.pause();
	}

	@Override
	protected void attachBaseContext(Context newBase) {
		super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
	}

    @Override
    public void registerFinishAddReceiver() {
        receiver = new BroadcastReceiver() {

            @Override
            public void onReceive(Context context, Intent intent) {
                String action = intent.getAction();
                String feedUrl = intent.getStringExtra(NetworkTaskManager.ADDED_FEED_URL);
                int errorReason = intent.getIntExtra(NetworkTaskManager.ADD_FEED_ERROR_REASON,
                        NetworkTaskManager.REASON_NOT_FOUND);
                presenter.handleFinish(action, feedUrl, errorReason);
            }
        };
        IntentFilter filter = new IntentFilter();
        filter.addAction(NetworkTaskManager.FINISH_UPDATE_ACTION);
        registerReceiver(receiver, filter);
    }

    @Override
    public void unregisterFinishAddReceiver() {
        if (receiver != null) {
            unregisterReceiver(receiver);
            receiver = null;
        }
    }

    @Override
    public void showInvalidUrlErrorToast() {
        showToastOnUiThread(R.string.add_rss_error_invalid_url, Toast.LENGTH_SHORT);
        GATrackerHelper.INSTANCE.sendEvent(getString(R.string.add_rss_from_intent_error));
    }

    @Override
    public void showGenericErrorToast() {
        showToastOnUiThread(R.string.add_rss_error_generic, Toast.LENGTH_SHORT);
        GATrackerHelper.INSTANCE.sendEvent(getString(R.string.add_rss_from_intent_error));
    }

    @Override
    public void finishView() {
        finish();
    }

    @Override
    public void showSuccessToast() {
        showToastOnUiThread(R.string.add_rss_success, Toast.LENGTH_SHORT);
    }

    @UiThread
    private void showToastOnUiThread(@StringRes final int res, final int toastLength) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(getApplicationContext(), res, toastLength).show();
            }
        });
    }
}

File: legacy/src/main/java/com/phicdy/mycuration/presentation/view/activity/FeedUrlHookActivity.kt
package com.phicdy.mycuration.presentation.view.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.annotation.UiThread
import androidx.appcompat.app.AppCompatActivity
import com.phicdy.mycuration.feature.util.changeTheme
import com.phicdy.mycuration.legacy.R
import com.phicdy.mycuration.presentation.presenter.FeedUrlHookPresenter
import com.phicdy.mycuration.presentation.view.FeedUrlHookView
import com.phicdy.mycuration.tracker.TrackerHelper
import io.github.inflationx.viewpump.ViewPumpContextWrapper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.koin.android.scope.currentScope
import org.koin.core.parameter.parametersOf
import kotlin.coroutines.CoroutineContext

class FeedUrlHookActivity : AppCompatActivity(), FeedUrlHookView, CoroutineScope {

    private val job = Job()
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    private val presenter: FeedUrlHookPresenter by currentScope.inject {
        parametersOf(
                this,
                if (intent.action == null) "" else intent.action,
                if (intent.dataString == null) "" else intent.dataString,
                if (intent.extras == null) "" else intent.extras?.getCharSequence(Intent.EXTRA_TEXT, "")
                        ?: "",
                this
        )
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_feed_url_hook)
        launch { presenter.create() }
    }

    override fun onResume() {
        super.onResume()
        changeTheme()
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(ViewPumpContextWrapper.wrap(newBase))
    }

    override fun showInvalidUrlErrorToast() {
        showToastOnUiThread(R.string.add_rss_error_invalid_url, Toast.LENGTH_SHORT)
    }

    override fun showGenericErrorToast() {
        showToastOnUiThread(R.string.add_rss_error_generic, Toast.LENGTH_SHORT)
    }

    override fun finishView() {
        finish()
    }

    override fun trackFailedUrl(url: String) {
        TrackerHelper.sendFailedParseUrl(getString(R.string.add_rss_from_intent_error), url)
    }

    override fun showSuccessToast() {
        showToastOnUiThread(R.string.add_rss_success, Toast.LENGTH_SHORT)
    }

    @UiThread
    private fun showToastOnUiThread(@StringRes res: Int, toastLength: Int) {
        runOnUiThread { Toast.makeText(applicationContext, res, toastLength).show() }
    }
}
